src/utils/bitfield.rs

```diff
@@ -0,0 +1,231 @@
+                                    range.start << range.start);
+                (self & mask) | (value << range.start)
+            }
+        })*
+    }
+}
+
+impl_bitfields!(for u8, u16, u32, u64);
+
+fn to_regular_range<T: RangeBounds<usize>>(range: &T, maximun: usize) -> Range<usize> {
+    let start = match range.start_bound() {
+        Bound::Excluded(&value) => value + 1,
+        Bound::Included(&value) => value,
+        Bound::Unbounded => 0,
+    };
+    let end = match range.end_bound() {
+        Bound::Excluded(&value) => value,
+        Bound::Included(&value) => value + 1,
+        Bound::Unbounded => maximun,
+    };
+
+    start..end
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // ############################################## get_bit
+
+    #[test_case]
+    fn get_bit_0_set_u16() {
+        let val: u16 = 1;
+        assert_eq!(val.get_bit(0), true);
+    }
+
+    #[test_case]
+    fn get_bit_0_unset_u8() {
+        let val: u8 = !0 - 1;
+        assert_eq!(val.get_bit(0), false);
+    }
+
+    #[test_case]
+    fn get_bit_15_set_u16() {
+        let val: u16 = 0x8000;
+        assert_eq!(val.get_bit(15), true);
+    }
+
+    // ############################################## get_bits
+
+    #[test_case]
+    fn get_bits_full_u16() {
+        let val: u16 = 0xdead;
+        assert_eq!(val.get_bits(0..16), 0xdead);
+    }
+
+    #[test_case]
+    fn get_bits_low_half_u32() {
+        let val: u32 = 0xdeadbeef;
+        assert_eq!(val.get_bits(0..16), 0xbeef);
+    }
+
+    #[test_case]
+    fn get_bits_high_half_u32() {
+        let val: u32 = 0xdeadbeef;
+        assert_eq!(val.get_bits(16..=31), 0xdead);
+    }
+
+    #[test_case]
+    fn get_bits_one_u8() {
+        let val: u8 = 0b10101010;
+        assert_eq!(val.get_bits(1..2), 1);
+    }
+
+    #[test_case]
+    fn get_bits_unbounded_u64() {
+        let val: u64 = 0xdeadbeefcafeface;
+        assert_eq!(val.get_bits(0..), val);
+    }
+
+    // ############################################## set_bit
+
+    #[test_case]
+    fn set_bit_0_u16() {
+        let val: u16 = 0;
+        assert_eq!(val.set_bit(0, true), 1);
+    }
+
+    #[test_case]
+    fn unset_bit_15_u16() {
+        let val: u16 = 0x8000;
+        assert_eq!(val.set_bit(15, false), 0);
+    }
+
+    // ############################################## set_bits
+
+    #[test_case]
+    fn set_bits_full_included_u16() {
+        let val: u16 = 0;
+        assert_eq!(val.set_bits(0..=15, 0xfeaf), 0xfeaf);
+    }
+
+    #[test_case]
+    fn set_bits_full_excluded_u16() {
+        let val: u16 = 0;
+        assert_eq!(val.set_bits(0..16, 0xfeaf), 0xfeaf);
+    }
+
+    #[test_case]
+    fn set_bits_full_unbounded_u32() {
+        let val: u32 = 0;
+        assert_eq!(val.set_bits(0.., 0xfeafbade), 0xfeafbade);
+    }
+
+    #[test_case]
+    fn unset_bits_u8() {
+        let val: u8 = 0b11111111;
+        assert_eq!(val.set_bits(4..=6, 0), 0b10001111);
+    }
+
```

Comment:
There is no example of panicking test, I'm not entirely sure but I think rust has some `#[should_panic]` modifier or something along this.

Reply:
If I understood things correctly, this feature is from the `test` crate which depends on `std`

Reply:
Yeah you're right. We might circumvent this by adding an integration test and modifying the panic handler.